File: mcp-client-bridge/src/reconnection-manager.test.ts
import { ReconnectionManager, BridgeState } from './reconnection-manager';
import { HttpStreamableClient } from './http-streamable-client';
import { BridgeConfig } from './bridge-config';

// Mock HttpStreamableClient
jest.mock('./http-streamable-client');

describe('ReconnectionManager', () => {
  let config: BridgeConfig;
  let mockClient: jest.Mocked<HttpStreamableClient>;

  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    config = {
      orchestratorUrls: ['http://localhost:8080'],
      orchestratorUrl: 'http://localhost:8080',
      reconnectEnabled: true,
      maxReconnectDelayMs: 100,
      baseReconnectDelayMs: 10,
      requestTimeoutMs: 30000,
      connectionTimeoutMs: 5000,
      maxRetryBeforeRotate: 3,
      enableLocalStreamWrite: true,
      enableLocalServers: true,
      configPath: undefined,
      healthIntervalMs: 30000,
      routingRefreshMs: 60000,
      pingIntervalMs: 30000,
      pingTimeoutMs: 5000,
      token: null,
    };
    mockClient = new HttpStreamableClient(config) as jest.Mocked<HttpStreamableClient>;
  });

  it('should start in DISCONNECTED state', () => {
    const manager = new ReconnectionManager(config, mockClient);
    expect(manager.state).toBe(BridgeState.DISCONNECTED);
  });

  it('should transition to CONNECTED on successful init', async () => {
    mockClient.initialize = jest.fn().mockResolvedValue(true);
    Object.defineProperty(mockClient, 'isConnected', { get: () => true });

    const manager = new ReconnectionManager(config, mockClient);
    const result = await manager.connectWithRetry();

    expect(result).toBe(true);
    expect(manager.state).toBe(BridgeState.CONNECTED);
  });

  it('should transition to DISCONNECTED after all URLs fail', async () => {
    mockClient.initialize = jest.fn().mockResolvedValue(false);
    Object.defineProperty(mockClient, 'isConnected', { get: () => false });

    const manager = new ReconnectionManager(config, mockClient);
    const result = await manager.connectWithRetry();

    expect(result).toBe(false);
    expect(manager.state).toBe(BridgeState.DISCONNECTED);
  });

  it('should try multiple URLs on failover', async () => {
    config.orchestratorUrls = ['http://primary:8080', 'http://backup:8080'];
    config.orchestratorUrl = 'http://primary:8080';

    mockClient.initialize = jest.fn()
      .mockResolvedValueOnce(false)
      .mockResolvedValueOnce(true);
    Object.defineProperty(mockClient, 'isConnected', { get: () => true });

    const manager = new ReconnectionManager(config, mockClient);
    const result = await manager.connectWithRetry();

    expect(result).toBe(true);
    expect(mockClient.initialize).toHaveBeenCalledTimes(2);
    expect(mockClient.initialize).toHaveBeenCalledWith('http://primary:8080');
    expect(mockClient.initialize).toHaveBeenCalledWith('http://backup:8080');
  });

  it('should not reconnect when disabled', async () => {
    config.reconnectEnabled = false;
    mockClient.initialize = jest.fn().mockResolvedValue(false);
    Object.defineProperty(mockClient, 'isConnected', { get: () => false });

    const manager = new ReconnectionManager(config, mockClient);
    await manager.reconnectLoop();

    expect(mockClient.initialize).not.toHaveBeenCalled();
  });
});

File: orchestrator-bridge/src/test/kotlin/com/orchestrator/mcp/bridge/ReconnectionManagerTest.kt
package com.orchestrator.mcp.bridge

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.*
import kotlinx.coroutines.test.runTest

class ReconnectionManagerTest : DescribeSpec({

    describe("ReconnectionManager") {

        it("should transition to CONNECTED on successful init") {
            runTest {
                val client = mockk<HttpStreamableClient>()
                coEvery { client.initialize(any(), any()) } returns true
                every { client.isConnected } returns true

                val config = BridgeConfig(
                    orchestratorUrls = listOf("http://localhost:8080"),
                    baseReconnectDelayMs = 10,
                    maxReconnectDelayMs = 100
                )
                val manager = ReconnectionManager(config, client)

                val result = manager.connectWithRetry()
                result shouldBe true
                manager.state shouldBe BridgeState.CONNECTED
            }
        }

        it("should transition to DISCONNECTED after all URLs fail") {
            runTest {
                val client = mockk<HttpStreamableClient>()
                coEvery { client.initialize(any(), any()) } returns false
                every { client.isConnected } returns false

                val config = BridgeConfig(
                    orchestratorUrls = listOf("http://localhost:8080"),
                    baseReconnectDelayMs = 1,
                    maxReconnectDelayMs = 10
                )
                val manager = ReconnectionManager(config, client)

                val result = manager.connectWithRetry()
                result shouldBe false
                manager.state shouldBe BridgeState.DISCONNECTED
            }
        }

        it("should start in DISCONNECTED state") {
            val client = mockk<HttpStreamableClient>()
            val config = BridgeConfig(orchestratorUrls = listOf("http://localhost:8080"))
            val manager = ReconnectionManager(config, client)
            manager.state shouldBe BridgeState.DISCONNECTED
        }

        it("should try multiple URLs on failover") {
            runTest {
                val client = mockk<HttpStreamableClient>()
                coEvery { client.initialize("http://primary:8080", any()) } returns false
                coEvery { client.initialize("http://backup:8080", any()) } returns true
                every { client.isConnected } returns true

                val config = BridgeConfig(
                    orchestratorUrls = listOf("http://primary:8080", "http://backup:8080"),
                    baseReconnectDelayMs = 1,
                    maxReconnectDelayMs = 10
                )
                val manager = ReconnectionManager(config, client)

                val result = manager.connectWithRetry()
                result shouldBe true
                manager.state shouldBe BridgeState.CONNECTED
            }
        }
    }
})
